napari/components/dims.py

```diff
@@ -140,54 +175,106 @@ def _check_dims(cls, values):
             )
 
         # Check the axis labels tuple has same number of elements as ndim
-        if len(values['axis_labels']) < ndim:
+        labels_ndim = len(values['axis_labels'])
+        if labels_ndim < ndim:
             # Append new "default" labels to existing ones
-            if values['axis_labels'] == tuple(
-                map(str, range(len(values['axis_labels'])))
-            ):
+            if values['axis_labels'] == tuple(map(str, range(labels_ndim))):
                 values['axis_labels'] = tuple(map(str, range(ndim)))
             else:
                 values['axis_labels'] = (
-                    tuple(map(str, range(ndim - len(values['axis_labels']))))
+                    tuple(map(str, range(ndim - labels_ndim)))
                     + values['axis_labels']
                 )
-        elif len(values['axis_labels']) > ndim:
+        elif labels_ndim > ndim:
             values['axis_labels'] = values['axis_labels'][-ndim:]
 
         return values
 
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        # to be removed if/when deprecating current_step
+        self.events.point_slider.connect(self.events.current_step)
+
     @property
-    def nsteps(self) -> Tuple[int, ...]:
-        """Tuple of int: Number of slider steps for each dimension."""
+    def nsteps(self) -> Tuple[float]:
         return tuple(
-            int((max_val - min_val) / step_size)
-            for min_val, max_val, step_size in self.range
+            # "or 1" ensures degenerate dimension works
+            int((rng.stop - rng.start) / (rng.step or 1))
+            for rng in self.range
         )
 
+    @nsteps.setter
+    def nsteps(self, value):
+        self.range = [
+            (rng.start, rng.stop, (rng.stop - rng.start) / nsteps)
+            for rng, nsteps in zip(self.range, value)
+        ]
```

Comment:
I'm super curious/concerned about off by one / off by a rounding error here. I'm getting vibes of #1686 (and I think other similar bugs, though I can't find the other references), because generally `np.arange(rng.start, rng.stop, rng.step)` will have either `nsteps` or `nsteps+1` elements depending on floating point rounding error here. Do you have a good grip on those issues here?